### PR TITLE
Remove MPI 2.x logic

### PR DESCRIPTION
**Pull request by pengfej**
An update on Issue #13447: Remove MPI 2.X logic

### CODE REVIEW

File: source/base/mpi.cc [hunk 1]
```diff
@@ -435,130 +363,8 @@ namespace Utilities
...
-        for (unsigned int j = 0; j < max_n_destinations; ++j)
-          if (all_destinations[i * max_n_destinations + j] == myid)
-            origins.push_back(i);
-          else if (all_destinations[i * max_n_destinations + j] ==
-                   numbers::invalid_unsigned_int)
-            break;
-
-      return origins;
-#  endif
+      // If that was not the case, we need to use the remainder of the code
```
Comment (tjhei) [inline source/base/mpi.cc:366]: these two lines need to be removed.

File: source/base/mpi.cc [hunk 2]
```diff
@@ -584,7 +390,6 @@ namespace Utilities
       // then we can short-cut the process using a consensus algorithm (which
       // is implemented only for the case of unique destinations, and also only
```
Comment (tjhei) [inline source/base/mpi.cc:391]: can you remove the text in parenthesis here

File: source/base/mpi.cc [hunk 3]
```diff
@@ -593,68 +398,24 @@ namespace Utilities
             .size();
         }
       else
-#  endif
-        {
```
Comment (tjhei) [inline source/base/mpi.cc:597]: this code block needs to stay, right?

File: source/base/mpi.cc [hunk 4]
```diff
@@ -1289,14 +1050,10 @@ namespace Utilities
       const int ierr = MPI_Barrier(comm);
       AssertThrowMPI(ierr);
 
-#  if 0 && DEAL_II_MPI_VERSION_GTE(3, 0)
```
Comment (tjhei) [inline source/base/mpi.cc:1292]: please keep this line and change it to 
```#  if 0```

File: source/base/mpi.cc [hunk 5]
```diff
@@ -1289,14 +1050,10 @@ namespace Utilities
...
       // wait for non-blocking barrier to finish. This is a noop the
       // first time we lock().
       const int ierr = MPI_Wait(&request, MPI_STATUS_IGNORE);
       AssertThrowMPI(ierr);
-#  else
```
Comment (tjhei) [inline source/base/mpi.cc:1297]: leave this code block

File: source/base/mpi.cc [hunk 6]
```diff
@@ -1320,13 +1077,8 @@ namespace Utilities
       // in the lock and unlock. It needs to be tested, if we can move
       // to a nonblocking barrier (code disabled below):
 
-#  if 0 && DEAL_II_MPI_VERSION_GTE(3, 0)
```
Comment (tjhei) [inline source/base/mpi.cc:1323]: same as above

File: source/base/mpi.cc [hunk 7]
```diff
@@ -1320,13 +1077,8 @@ namespace Utilities
...
       const int ierr = MPI_Ibarrier(comm, &request);
       AssertThrowMPI(ierr);
-#  else
```
Comment (tjhei) [inline source/base/mpi.cc:1326]: keep this

File: tests/mpi/create_mpi_datatype_01.cc [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// ---------------------------------------------------------------------
+// // ---------------------------------------------------------------------
```
Comment (tjhei) [inline tests/mpi/create_mpi_datatype_01.cc:1]: please undo this accidental change

**Review by tjhei — COMMENTED** [review]
almost, I commented in a few places that seem incorrect

File: source/base/mpi.cc (hunk 3 above)
Comment (pengfej) [reply source/base/mpi.cc:597]: I checked the original code and my code for a few times, and it seems like I didn't remove this block, but instead I added a {} after the else, that might confuse the system, I'm gonna try to remove the {} and see if it helps.

File: include/deal.II/base/mpi.h [hunk 9]
```diff
@@ -80,19 +80,7 @@ using MPI_Op       = int;
...
-#  else
-
-#    include <type_traits>
-
-#    define DEAL_II_MPI_CONST_CAST(expr)     \
-      const_cast<typename std::remove_const< \
-        typename std::remove_pointer<decltype(expr)>::type>::type *>(expr)
-
-#  endif
+#  define DEAL_II_MPI_CONST_CAST(expr) (expr)
```
Comment (bangerth) [inline include/deal.II/base/mpi.h:83]: This means that (in a follow-up) we can get rid of all places where this macro is used!

File: source/base/mpi.cc [hunk 10]
```diff
@@ -183,76 +183,7 @@ namespace Utilities
                  const int        tag,
                  MPI_Comm *       new_comm)
     {
-#  if DEAL_II_MPI_VERSION_GTE(3, 0)
       return MPI_Comm_create_group(comm, group, tag, new_comm);
-#  else
```
Comment (bangerth) [inline source/base/mpi.cc:188]: And this means that we can get rid of this function as well. (Also as a follow-up.)

File: include/deal.II/base/mpi.h (hunk 9 above)
Comment (tjhei) [reply include/deal.II/base/mpi.h:83]: Good point. I let Pengfei do that as well... 😉

File: source/base/mpi.cc [hunk 11]
```diff
@@ -580,11 +384,6 @@ namespace Utilities
         return (my_destinations.size() == n_destinations);
       }();
 
-      // If all processes report that they have unique destinations,
-      // then we can short-cut the process using a consensus algorithm (which
```
Comment (tjhei) [inline source/base/mpi.cc:584]: Please leave these lines untouched and just remove the text in parentheses.

Comment (pengfej) [reply source/base/mpi.cc:584]: Done! I will merge all commits when we are done!

File: source/base/mpi.cc [hunk 12]
```diff
@@ -435,130 +363,6 @@ namespace Utilities
...
-
-      if (destinations.size() > 0)
-        {
-          const int ierr = MPI_Waitall(destinations.size(),
-                                       send_requests.data(),
-                                       MPI_STATUSES_IGNORE);
-          AssertThrowMPI(ierr);
-        }
-
-      return origins;
```
Comment (marcfehling) [inline source/base/mpi.cc:508]: You still need to keep this part.

File: source/base/mpi.cc [hunk 13]
```diff
@@ -611,49 +412,6 @@ namespace Utilities
...
-          const int ierr = MPI_Reduce_scatter_block(dest_vector.data(),
-                                                    &n_recv_from,
-                                                    1,
-                                                    MPI_UNSIGNED,
-                                                    MPI_SUM,
-                                                    mpi_comm);
-
-          AssertThrowMPI(ierr);
-
-          return n_recv_from;
```
Comment (tjhei) [inline source/base/mpi.cc:629]: keep these lines 616-629

File: source/base/mpi.cc (hunk 12 above)
Comment (pengfej) [reply source/base/mpi.cc:508]: you mean keeping line 447-508?

File: source/base/mpi.cc (hunk 13 above)
Comment (pengfej) [reply source/base/mpi.cc]: Should be good now!

File: source/base/mpi.cc (hunk 12 above)
Comment (tjhei) [reply source/base/mpi.cc:508]: yes

Comment (pengfej) [reply source/base/mpi.cc]: should be good by now

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Looks good to me. OK to merge once the tests all succeed?

**Comment by tjhei** [comment]
Pengfei, can you please squash all commits into a single one?

**Review by tjhei — APPROVED** [review]
Looks good to go to me.

**Comment by pengfej** [comment]
> Pengfei, can you please squash all commits into a single one?

Just did!

File: source/base/mpi.cc [hunk 14]
```diff
@@ -435,16 +363,11 @@ namespace Utilities
...
 
-        // So we need to run a different algorithm, specifically one that
-        // requires more memory -- MPI_Reduce_scatter_block will require memory
-        // proportional to the number of processes involved; that function is
-        // also only available for MPI 2.2 or later:
-#  if DEAL_II_MPI_VERSION_GTE(2, 2)
+      // So we need to run a different algorithm, specifically one that
+      // requires more memory -- MPI_Reduce_scatter_block will require memory
+      // proportional to the number of processes involved; that function is
+      // also only available for MPI 2.2 or later:
```
Comment (marcfehling) [inline source/base/mpi.cc:370]:
```suggestion
      // proportional to the number of processes involved:
```

File: source/base/mpi.cc [hunk 15]
```diff
@@ -581,10 +451,8 @@ namespace Utilities
       }();
 
       // If all processes report that they have unique destinations,
-      // then we can short-cut the process using a consensus algorithm (which
-      // is implemented only for the case of unique destinations, and also only
-      // for MPI 3 and later):
-#  if DEAL_II_MPI_VERSION_GTE(3, 0)
+      // then we can short-cut the process using a consensus algorithm:
+
```
Comment (marcfehling) [inline source/base/mpi.cc:455]:
```suggestion
      // then we can short-cut the process using a consensus algorithm
      // (which is implemented only for the case of unique destinations):
```
I found this part of the parentheses still helpful

**Review by marcfehling — APPROVED** [review]
Looks good to me! I just have two comments on the documentation in the latest iteration.

**Comment by masterleinad** [comment]
```
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc: In constructor ‘Euler_DG::EulerOperator<dim, degree, n_points_1d>::EulerOperator(dealii::TimerOutput&)’:
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:522:29: error: ‘MPI_COMM_TYPE_SHARED’ was not declared in this scope
                             MPI_COMM_TYPE_SHARED,
                             ^~~~~~~~~~~~~~~~~~~~
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:522:29: note: suggested alternative: ‘MPI_COMM_WORLD’
                             MPI_COMM_TYPE_SHARED,
                             ^~~~~~~~~~~~~~~~~~~~
                             MPI_COMM_WORLD
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:524:29: error: ‘MPI_INFO_NULL’ was not declared in this scope
                             MPI_INFO_NULL,
                             ^~~~~~~~~~~~~
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:524:29: note: suggested alternative: ‘MPI_COMM_NULL’
                             MPI_INFO_NULL,
                             ^~~~~~~~~~~~~
                             MPI_COMM_NULL
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:521:9: error: there are no arguments to ‘MPI_Comm_split_type’ that depend on a template parameter, so a declaration of ‘MPI_Comm_split_type’ must be available [-fpermissive]
         MPI_Comm_split_type(MPI_COMM_WORLD,
         ^~~~~~~~~~~~~~~~~~~
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:521:9: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc: In instantiation of ‘Euler_DG::EulerOperator<dim, degree, n_points_1d>::EulerOperator(dealii::TimerOutput&) [with int dim = 2; int degree = 5; int n_points_1d = 7]’:
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:1343:18:   required from ‘Euler_DG::EulerProblem<dim>::EulerProblem() [with int dim = 2]’
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:1605:31:   required from here
/home/runner/work/dealii/dealii/examples/step-76/step-76.cc:521:28: error: ‘MPI_Comm_split_type’ was not declared in this scope
         MPI_Comm_split_type(MPI_COMM_WORLD,
         ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
                             MPI_COMM_TYPE_SHARED,
                             ~~~~~~~~~~~~~~~~~~~~~
                             rank,
                             ~~~~~
                             MPI_INFO_NULL,
                             ~~~~~~~~~~~~~~
                             &subcommunicator);
```

File: examples/step-76/step-76.cc [hunk 16]
```diff
@@ -510,7 +510,6 @@ namespace Euler_DG
   EulerOperator<dim, degree, n_points_1d>::EulerOperator(TimerOutput &timer)
     : timer(timer)
   {
-#if DEAL_II_MPI_VERSION_GTE(3, 0)
```
Comment (tjhei) [inline examples/step-76/step-76.cc:513]: This causes the bug. Can you undo all changes in step-76? We can address this one separately.

**Review by masterleinad — APPROVED** [review]

**Comment by tjhei** [comment]
thank you, @pengfej !

**Pull request merged**